ci2/ci/github.py

```diff
@@ -128,17 +165,78 @@ def from_gh_json(gh_json, target_branch):
 
         review_state = 'pending'
         for login, state in latest_state_by_login.items():
-            if (state == 'CHANGES_REQUESTED'):
+            if state == 'CHANGES_REQUESTED':
                 review_state = 'changes_requested'
                 break
-            elif (state == 'APPROVED'):
+            if state == 'APPROVED':
                 review_state = 'approved'
             else:
-                assert state == 'DISMISSED' or state == 'COMMENTED', state
+                assert state in ('DISMISSED', 'COMMENTED'), state
 
         self.review_state = review_state
 
+    async def start_build(self, batch_client):
+        assert not self.batch
+
+        # FIXME this needs to be per-PR and async
+        try:
+            async with repos_lock:
+                target_repo = self.target_branch.branch.repo
+                repo_dir = f'repos/{target_repo.short_str()}'
+
+                shell('git', 'config', 'user.email', 'hail-ci-leader@example.com')
+                shell('git', 'config', 'user.name', 'hail-ci-leader')
+
+                if not os.path.isdir(repo_dir):
+                    os.makedirs(repo_dir, exist_ok=True)
+                    shell('git', '-C', repo_dir, 'clone', self.target_branch.branch.repo.url, '.')
+
+                if sp.run([
+                        '/bin/sh', '-c',
+                        f'git -C {shq(repo_dir)} remote | grep -q {shq(self.source_repo.short_str())}'
+                ]).returncode != 0:
+                    shell('git', '-C', repo_dir, 'remote', 'add', self.source_repo.short_str(), self.source_repo.url)
+
+                shell('git', '-C', repo_dir, 'reset', '--merge')
+                shell('git', '-C', repo_dir, 'fetch', 'origin')
+                shell('git', '-C', repo_dir, 'fetch', self.source_repo.short_str())
+                shell('git', '-C', repo_dir, 'checkout', self.target_branch.sha)
+                shell('git', '-C', repo_dir, 'merge', self.source_sha, '-m', 'merge PR')
+
+                self.sha = (sp.check_output(['git', 'rev-parse', 'HEAD'])
+                            .decode('utf-8')
+                            .strip())
```

Comment:
we can change `shell` to return the result of `sp.run` in which case this could be 
```
shell('git', 'rev-parse', 'HEAD').stdout.decode('utf-8').strip()
```